File: assignment6/src/com/northeastern/edu/question4.java
package com.northeastern.edu;

import java.util.Deque;
import java.util.LinkedList;

public class question4 {


    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    public class Codec {
        String NULL = "#";
        String SEP = ",";
        // Encodes a tree to a single string.
        public String serialize(TreeNode root) {
            StringBuilder sb = new StringBuilder();
            return serializeTree(root, sb).toString();
        }

        public StringBuilder serializeTree(TreeNode root, StringBuilder sb) {
            if (root == null) {
                sb.append(NULL).append(SEP);
                return sb;
            }
            sb.append(root.val).append(SEP);
            serializeTree(root.left, sb);
            serializeTree(root.right, sb);
            return sb;
        }

        // Decodes your encoded data to tree.
        public TreeNode deserialize(String data) {
            Deque<String> deque = new LinkedList<>();
            String[] tmp = data.split(",");
            for (String ss : tmp) {
                deque.offer(ss);
            }
            return constructTree(deque);
        }

        public TreeNode constructTree(Deque<String> deque) {
            if (deque.isEmpty()) {
                return null;
            }
            String tmp = deque.poll();
            if (tmp.equals(NULL)) {
                return null;
            }
            TreeNode root = new TreeNode(Integer.parseInt(tmp));
            root.left = constructTree(deque);
            root.right = constructTree(deque);
            return root;
        }
    }

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

//    public static void main(String[] args) {
//        TreeNode node1 = new TreeNode(4);
//        TreeNode node2 = new TreeNode(3);
//        TreeNode node3 = new TreeNode(5);
//        TreeNode node4 = new TreeNode(2);
//        TreeNode node5 = new TreeNode(0);
//        TreeNode node6 = new TreeNode(8);
//        TreeNode node7 = new TreeNode(7);
//        TreeNode node8 = new TreeNode(1);
//        node1.left = node2;
//        node2.left = node3;
//        node3.right = node4;
//        node2.right = node8;
//        node3.left = node6;
//        node6.left = node7;
//        node6.right = node5;
//
//        Codec ser = new Codec();
//        Codec deser = new Codec();
//        TreeNode ans = deser.deserialize(ser.serialize(node1));
//    }
}

File: final/question4.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val 
        self.left = left
        self.right = right

class solution:
    def lowestCommonAncestor(self, root, p, q):
        def check(root, p, q):
            if (not root):
                return None
            if (root.val == p.val or root.val == q.val):
                return root

            leftVal = check(root.left, p, q)
            rightVal = check(root.right, p, q)

            if (leftVal and rightVal):
                return root
            elif (leftVal):
                return leftVal
            else:
                return rightVal

        return check(root, p, q)
    
obj = solution()
node1 = TreeNode(0)
node2 = TreeNode(2)
node3 = TreeNode(9)
node4 = TreeNode(10)
node5 = TreeNode(11)
node6 = TreeNode(4)
'''

    1
  /   \ 
2      9
  \   / \ 
  10 4   11

'''
node1.left = node2
node1.right = node3
node2.right = node4 
node3.left = node6 
node3.right = node5
# res = obj.lowestCommonAncestor(node1, node5, node6)
res = obj.lowestCommonAncestor(node1, node2, node3)
print(res.val)

    
